core/internal/wboperation/wboperation.go

```diff
@@ -0,0 +1,348 @@
+// Package wboperation is used to track the progress and status of async tasks
+// to provide feedback to the wandb user.
+//
+// The core abstraction is WandbOperation which represents a high level task,
+// like an artifact upload. Operations can be broken into parallel subtasks,
+// such as the multiple file uploads that are part of saving an artifact.
+//
+// A WandbOperation has an optional error status and progress indicator.
+// The error status conveys the most recent problem encountered by the
+// operation and the resulting operation status, like
+// "retrying HTTP 500 error".
+//
+// Operation-aware code should accept a `*WandbOperation` parameter
+// unless it already accepts a `context.Context`. If it has a `Context`,
+// it should use the operation in it, if any.
+//
+// Nil values are allowed and are treated like no-ops. Code should never
+// check whether *WandbOperation or *WandbOperations is nil.
+package wboperation
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+)
+
+type contextKey int
+
+const operationContextKey contextKey = iota
+const maxOperationsToReturn = 10
+
+// Get returns the operation in the context, or nil if there isn't one.
+func Get(ctx context.Context) *WandbOperation {
+	if ctx == nil {
+		return nil
+	}
+
+	if value, ok := ctx.Value(operationContextKey).(*WandbOperation); ok {
+		return value
+	}
+
+	return nil
+}
+
+// WandbOperations tracks the status and progress of ongoing operations
+// in the wandb internal process.
+//
+// A nil pointer acts like a no-op, as if operation tracking is disabled.
+type WandbOperations struct {
+	// mu is locked when creating or removing operations or subtasks.
+	mu sync.Mutex
+
+	// operations is all operations sorted in ascending order by startTime.
+	//
+	// We use a doubly-linked list because operations are removed frequently
+	// and in random order, and there may be a large number of them.
+	operations list.List
+}
+
+func NewOperations() *WandbOperations {
+	ops := &WandbOperations{}
+	ops.operations.Init()
+	return ops
+}
+
+// New starts a new operation with the given name.
+//
+// It is important to call Finish() on the operation once it is done.
+//
+// The description should be a present continuous verb phrase like
+// "uploading artifact my-dataset". It should not start with a capital
+// letter or end in punctuation.
```

Comment:
do we want to enforce it with some linter? (i didn't look for anything, but probably doable)
very low priority ofc

Reply:
It might not be hard to write one using the go/analysis package, but I don't think it's worth it. The docs spell it out clearly and it's easy to quickly audit all uses by looking at the call hierarchy.

I might make this accept a custom type and define all messages in one file.